src/MosaicViews.jl

```diff
@@ -366,40 +391,57 @@ function _padded_cat(imgs; center, fillvalue, dims)
         @warn msg
     end
     T = _filltype(imgs)
-    has_offsets = any(Base.has_offset_axes.(imgs))
-    if !has_offsets && length(unique(map(axes, imgs))) == 1
-        return of_eltype(T, cat(imgs...; dims=dims))
+    has_offsets = false
+    for img in imgs
+        has_offsets = Base.has_offset_axes(img)
+        has_offsets && break
+    end
+    if !has_offsets && has_common_axes(imgs)
+        return Base._cat_t(dims, T, imgs...)
     else
         if !has_offsets && !center
             # in this case the index of PaddedView starts from 1
             # hence we can directly pass them into `cat`
-            return of_eltype(T, cat(paddedviews(fillvalue, imgs...)...; dims=dims))
+            return Base._cat_t(dims, T, pv(imgs)...)
         else
-            # TODO: it's unidentified but this version allocates more memory
-            #       and become slower (~1.5X) than a direct `cat`
-            pad_fn = center ? sym_paddedviews : paddedviews
-            return of_eltype(T, reduce(pad_fn(fillvalue, imgs...)) do x, y
-                x = OffsetArray(x, 1 .- first.(axes(x)))
-                y = OffsetArray(y, 1 .- first.(axes(y)))
-                cat(x, y; dims=dims)
-            end)
+            if center
+                return Base._cat_t(dims, T, map(OffsetArrays.no_offset_view, sym_pv(imgs))...)
+            end
+            return Base._cat_t(dims, T, map(OffsetArrays.no_offset_view, pv(imgs))...)
         end
     end
 end
 
-function _filltype(As)
-    Ts = (typeof.(first.(As))..., eltype.(typeof(As).types)...)
-    C = foldl(PaddedViews.filltype, filter(x-> x!== Any, Ts))
-    T = promote_type(eltype.(Ts)...) # eltype(eltype(RGB{Float32})) == Float32
-    isconcretetype(T) ? PaddedViews.filltype(C, T) : C
+has_common_axes(@nospecialize(imgs)) = isempty(imgs) || all(isequal(axes(first(imgs))) ∘ axes, imgs)
+
+
+_gettype(A::AbstractArray{T}) where T = T === Any ? typeof(first(A)) : T
+
+# This uses Union{} as a sentinel eltype (all other types "beat" it),
+# and Bool as a near-neutral fill type.
+_filltype(As) = PaddedViews.filltype(Bool, _filltypeT(Union{}, As...))
+@inline _filltypeT(::Type{T}, A, tail...) where T = _filltypeT(promote_type_withcolor(T, _gettype(A)), tail...)
+_filltypeT(::Type{T}) where T = T
+
+# When the inputs are homogenous we can circumvent varargs despecialization
+# This also handles the case of empty `As` but concrete `T`.
+function _filltype(As::AbstractVector{A}) where A<:AbstractArray{T} where T
+    (!@isdefined(T) || T === Any) && return invoke(_filltype, Tuple{Any}, As)
```

Comment:
In some cases it can be inferred as `Union`. I think it actually came up in practice, but I don't remember the details. Probably worth removing it and trying to catch an error, then make sure that's supplied as a test case before adding it back.